src/lab16.py

```diff
@@ -233,26 +233,34 @@ def __init__(self, node, frame):
 
         self.descendants = ProtectedField(node, "descendants")
 
+    def layout_needed(self):
+        if self.zoom.dirty: return True
+        if self.width.dirty: return True
+        if self.height.dirty: return True
+        if self.x.dirty: return True
+        if self.y.dirty: return True
+        if self.descendants.dirty: return True
+        return False
+
     def layout(self, width, zoom):
         self.zoom.set(zoom)
+        self.width.set(width - 2 * device_px(HSTEP, zoom))
```

Comment:
Why put these two lines before the layout_needed check?

Reply:
Well, if the `zoom` or `width` argument has changed, we need to update those fields. The `layout_needed` method will check their dirty bits, but not diff the current value with the argument. We could change how zooming / resizing works but I think this is easier.

Reply:
If `layout_needed` checks `self.zoom.dirty` and `self.width.dirty`, then i'll always recompute layout if they change, right?

Reply:
I think I figured out the confusion. `zoom` and `width` aren't modified or marked by the `Frame` when the zoom level or frame width changes. (So they're not dirty until the `set` operation.) Would you rather do that—mark those fields dirty when the inputs change?

Reply:
That would be more consistent, so let's try that.

Reply:
Done. Honestly, I think this approach is a little worse, but take a look and let me know. It's not a huge difference either way.